README.txt
========
toposort
========

Overview
========

Implements a topological sort algorithm.

From `Wikipedia <http://en.wikipedia.org/wiki/Topological_sorting>`_:
In computer science, a topological sort (sometimes abbreviated topsort
or toposort) or topological ordering of a directed graph is a linear
ordering of its vertices such that for every directed edge uv from
vertex u to vertex v, u comes before v in the ordering.

Input data description
======================

The input to the toposort function is a dict describing the
dependencies among the input nodes. Each key is a dependent node, the
corresponding value is a set containing the dependent nodes.

Note that toposort does not care what the input node values mean: it
just compares them for equality. The examples here usually use
integers, but they could be any hashable type.

Typical usage
=============

The interpretation of the input data here is: If 2 depends on 11; 9
depends on 11, 8 and 10; 10 depends on 11 and 3 (and so on), then in what
order should we process the items such that all nodes are processed
before any of their dependencies?::

    >>> from toposort import toposort, toposort_flatten
    >>> list(toposort({2: {11},
    ...                9: {11, 8, 10},
    ...                10: {11, 3},
    ...                11: {7, 5},
    ...                8: {7, 3},
    ...               }))
    [{3, 5, 7}, {8, 11}, {2, 10}, {9}]

And the answer is: process 3, 5, and 7 (in any order); then process 8
and 11; then process 2 and 10; then process 9. Note that 3, 5, and 7
are returned first because they do not depend on anything. They are
then removed from consideration, and then 8 and 11 don't depend on
anything remaining. This process continues until all nodes are
returned, or a circular dependency is detected.

Circular dependencies
=====================

A circular dependency will raise a CyclicDependencyError, which is
derived from ValueError.  Here 1 depends on 2, and 2 depends on 1::

    >>> list(toposort({1: {2},
    ...                2: {1},
    ...               }))
    Traceback (most recent call last):
        ...
    toposort.CircularDependencyError: Circular dependencies exist among these items: {1:{2}, 2:{1}}

In addition, the 'data' attribute of the raised CyclicDependencyError
will contain a dict containing the subset of the input data involved
in the circular dependency.


Module contents
===============

``toposort(data)``

Returns an iterator describing the dependencies among nodes in the
input data. Each returned item will be a set. Each member of this set
has no dependencies in this set, or in any set previously returned.

``toposort_flatten(data, sort=True)``

Like toposort(data), except that it returns a list of all of the
depend values, in order. If sort is true, the returned nodes are sorted within
each group before they are appended to the result::

    >>> toposort_flatten({2: {11},
    ...                   9: {11, 8, 10},
    ...                   10: {11, 3},
    ...                   11: {7, 5},
    ...                   8: {7, 3},
    ...                  })
    [3, 5, 7, 8, 11, 2, 10, 9]

Note that this result is the same as the first example: ``[{3, 5, 7}, {8, 11}, {2, 10}, {9}]``,
except that the result is flattened, and within each set the nodes
are sorted.


Testing
=======

To test, run 'python setup.py test'. On python >= 3.0, this also runs the doctests.


CHANGES.txt
Change log
==========

1.5 2016-10-24 Eric V. Smith
----------------------------

* When a circular dependency error is detected, raise a specific
  exception, CircularDependencyError, which is a subclass of
  ValueError.  The 'data' attribute of the exception will contain the
  data involved in the circular dependency (issue #2).  Thanks
  lilydjwg for the initial patch.

* To make building wheels easier, always require setuptools in
  setup.py (issue #5).

* Mark wheel as being universal, that is, supporting both Python 2.7
  and 3.x (issue #7).

1.4 2015-05-16 Eric V. Smith
----------------------------

* Removed 'test' package, so it won't get installed by bdist_*. It's still
  included in sdists.

* No code changes.

1.3 2015-05-15 Eric V. Smith
----------------------------

* Fixed change log date.

* No code changes.

1.2 2015-05-15 Eric V. Smith
----------------------------

* Changed RPM name to python3-toposort if running with python 3.

* No code changes.

1.1 2014-07-24 Eric V. Smith
----------------------------

* Release version 1.1. No code changes.

* Add a README.txt entry on running the test suite.

* Fix missing test/__init__.py in the sdist.

1.0 2014-03-14 Eric V. Smith
----------------------------

* Release version 1.0. The API is stable.

* Add MANIFEST.in to MANIFEST.in, so that it is created in the sdist
  (issue #1).

0.2 2014-02-11 Eric V. Smith
----------------------------

* Modify setup.py to produce a RPM name of python-toposort for bdist_rpm.

0.1 2014-02-10 Eric V. Smith
----------------------------

* Initial release.


test/test_toposort.py
#######################################################################
# Tests for toposort module.
#
# Copyright 2014 True Blade Systems, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Notes:
#
########################################################################

import unittest

from toposort import toposort, toposort_flatten, CircularDependencyError

class TestCase(unittest.TestCase):
    def test_simple(self):
        self.assertEqual(list(toposort({2: {11},
                                        9: {11, 8},
                                        10: {11, 3},
                                        11: {7, 5},
                                        8: {7, 3},
                                        })),
                         [{3, 5, 7},
                          {8, 11},
                          {2, 9, 10},
                          ])

        # make sure self dependencies are ignored
        self.assertEqual(list(toposort({2: {2, 11},
                                        9: {11, 8},
                                        10: {10, 11, 3},
                                        11: {7, 5},
                                        8: {7, 3},
                                        })),
                         [{3, 5, 7},
                          {8, 11},
                          {2, 9, 10},
                          ])

        self.assertEqual(list(toposort({1: set()})),
                         [{1}])
        self.assertEqual(list(toposort({1: {1}})),
                         [{1}])

    def test_no_dependencies(self):
        self.assertEqual(list(toposort({1: {2},
                                        3: {4},
                                        5: {6},
                                        })),
                         [{2, 4, 6}, {1, 3, 5}])

        self.assertEqual(list(toposort({1: set(),
                                        3: set(),
                                        5: set(),
                                        })),
                         [{1, 3, 5}])

    def test_empty(self):
        self.assertEqual(list(toposort({})),
                         [])

    def test_strings(self):
        self.assertEqual(list(toposort({'2': {'11'},
                                        '9': {'11', '8'},
                                        '10': {'11', '3'},
                                        '11': {'7', '5'},
                                        '8': {'7', '3'},
                                        })),
                         [{'3', '5', '7'},
                          {'8', '11'},
                          {'2', '9', '10'},
                          ])

    def test_objects(self):
        o2 = object()
        o3 = object()
        o5 = object()
        o7 = object()
        o8 = object()
        o9 = object()
        o10 = object()
        o11 = object()
        self.assertEqual(list(toposort({o2: {o11},
                                        o9: {o11, o8},
                                        o10: {o11, o3},
                                        o11: {o7, o5},
                                        o8: {o7, o3, o8},
                                        })),
                         [{o3, o5, o7},
                          {o8, o11},
                          {o2, o9, o10},
                          ])

    def test_cycle(self):
        # a simple, 2 element cycle
        # make sure we can catch this both as ValueError and CircularDependencyError
        self.assertRaises(ValueError, list, toposort({1: {2},
                                                      2: {1}
                                                      }))
        with self.assertRaises(CircularDependencyError) as ex:
            list(toposort({1: {2},
                           2: {1}
                           }))
        self.assertEqual(ex.exception.data, {1: {2}, 2: {1}})

        # an indirect cycle
        self.assertRaises(ValueError, list, toposort({1: {2},
                                                      2: {3},
                                                      3: {1},
                                                      }))
        with self.assertRaises(CircularDependencyError) as ex:
            list(toposort({1: {2},
                           2: {3},
                           3: {1},
                           }))
        self.assertEqual(ex.exception.data, {1: {2}, 2: {3}, 3: {1}})

        # not all elements involved in a cycle
        with self.assertRaises(CircularDependencyError) as ex:
            list(toposort({1: {2},
                           2: {3},
                           3: {1},
                           5: {4},
                           4: {6},
                           }))
        self.assertEqual(ex.exception.data, {1: set([2]), 2: set([3]), 3: set([1])})

    def test_input_not_modified(self):
        data = {2: {11},
                9: {11, 8},
                10: {11, 3},
                11: {7, 5},
                8: {7, 3, 8},  # includes something self-referential
                }
        orig = data.copy()
        results = list(toposort(data))
        self.assertEqual(data, orig)

    def test_input_not_modified_when_cycle_error(self):
        data = {1: {2},
                2: {1},
                3: {4},
                }
        orig = data.copy()
        self.assertRaises(ValueError, list, toposort(data))
        self.assertEqual(data, orig)


class TestCaseAll(unittest.TestCase):
    def test_sort_flatten(self):
        data = {2: {11},
                9: {11, 8},
                10: {11, 3},
                11: {7, 5},
                8: {7, 3, 8},  # includes something self-referential
                }
        expected = [{3, 5, 7}, {8, 11}, {2, 9, 10}]
        self.assertEqual(list(toposort(data)), expected)

        # now check the sorted results
        results = []
        for item in expected:
            results.extend(sorted(item))
        self.assertEqual(toposort_flatten(data), results)

        # and the unsorted results. break the results up into groups to compare them
        actual = toposort_flatten(data, False)
        results = [{i for i in actual[0:3]}, {i for i in actual[3:5]}, {i for i in actual[5:8]}]
        self.assertEqual(results, expected)


class TestAll(unittest.TestCase):
    def test_all(self):
        import toposort

        # check that __all__ in the module contains everything that should be
        #  public, and only those symbols
        all = set(toposort.__all__)

        # check that things in __all__ only appear once
        self.assertEqual(len(all), len(toposort.__all__),
                         'some symbols appear more than once in __all__')

        # get the list of public symbols
        found = set(name for name in dir(toposort) if not name.startswith('_'))

        # make sure it matches __all__
        self.assertEqual(all, found)

unittest.main()


toposort.py
#######################################################################
# Implements a topological sort algorithm.
#
# Copyright 2014 True Blade Systems, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Notes:
#  Based on http://code.activestate.com/recipes/578272-topological-sort
#   with these major changes:
#    Added unittests.
#    Deleted doctests (maybe not the best idea in the world, but it cleans
#     up the docstring).
#    Moved functools import to the top of the file.
#    Changed assert to a ValueError.
#    Changed iter[items|keys] to [items|keys], for python 3
#     compatibility. I don't think it matters for python 2 these are
#     now lists instead of iterables.
#    Copy the input so as to leave it unmodified.
#    Renamed function from toposort2 to toposort.
#    Handle empty input.
#    Switch tests to use set literals.
#
########################################################################

from functools import reduce as _reduce

__all__ = ['toposort', 'toposort_flatten']

class CyclicDependency(ValueError):
    def __init__(self, cyclic):
        s = 'Cyclic dependencies exist among these items: {}'.format(', '.join(repr(x) for x in cyclic.items()))
        super(CyclicDependency, self).__init__(s)
        self.cyclic = cyclic

def toposort(data):
    """Dependencies are expressed as a dictionary whose keys are items
and whose values are a set of dependent items. Output is a list of
sets in topological order. The first set consists of items with no
dependences, each subsequent set consists of items that depend upon
items in the preceeding sets.
"""

    # Special case empty input.
    if len(data) == 0:
        return

    # Copy the input so as to leave it unmodified.
    data = data.copy()

    # Ignore self dependencies.
    for k, v in data.items():
        v.discard(k)
    # Find all items that don't depend on anything.
    extra_items_in_deps = _reduce(set.union, data.values()) - set(data.keys())
    # Add empty dependences where needed.
    data.update({item:set() for item in extra_items_in_deps})
    while True:
        ordered = set(item for item, dep in data.items() if len(dep) == 0)
        if not ordered:
            break
        yield ordered
        data = {item: (dep - ordered)
                for item, dep in data.items()
                    if item not in ordered}
    if len(data) != 0:
        raise CyclicDependency(data)


def toposort_flatten(data, sort=True):
    """Returns a single list of dependencies. For any set returned by
toposort(), those items are sorted and appended to the result (just to
make the results deterministic)."""

    result = []
    for d in toposort(data):
        result.extend((sorted if sort else list)(d))
    return result


setup_base.py
from __future__ import print_function
from distutils.core import setup, Command

# run our tests
class PyTest(Command):
    user_options = []
    def initialize_options(self):
        pass
    def finalize_options(self):
        pass
    def run(self):
        import sys, subprocess
        tests = [('test suite', ['-m', 'test.test_toposort']),
                 ]
        if sys.hexversion >= 0x03000000:
            # Skip doctests for python < 3.0. They use set literal reprs, which
            #  are different in 2.7. Testing under 3.x is good enough.
            tests.append(('doctests',   ['-m' 'doctest', 'README.txt']))
        for name, cmds in tests:
            print(name)
            errno = subprocess.call([sys.executable] + cmds)
            if errno != 0:
                raise SystemExit(errno)
        print('test complete')


def base_setup(name):
    setup(name=name,
          version='0.1',
          url='https://bitbucket.org/ericvsmith/toposort',
          author='Eric V. Smith',
          author_email='eric@trueblade.com',
          description='Implements a topological sort algorithm.',
          long_description=open('README.txt').read() + '\n' + open('CHANGES.txt').read(),
          classifiers=['Development Status :: 5 - Production/Stable',
                       'Intended Audience :: Developers',
                       'License :: OSI Approved :: Apache Software License',
                       'Topic :: Software Development :: Libraries :: Python Modules',
                       'Programming Language :: Python :: 2.7',
                       'Programming Language :: Python :: 3.3',
                       ],
          license='Apache License Version 2.0',
          py_modules=['toposort'],

          cmdclass = {'test': PyTest},
          )
